src/crimson/os/seastore/seastore.cc

```diff
@@ -760,6 +769,18 @@ SeaStore::tm_ret SeaStore::_setattrs(
 {
   logger().debug("{} onode={}",
                 __func__, *onode);
+  auto layout = onode->get_mutable_layout(*ctx.transaction);
+  auto pattr = aset.find(OI_ATTR);
+  if (pattr != aset.end()) {
+    if (__builtin_expect(
+        pattr->second.length() > onode_layout_t::MAX_OI_LENGTH,
+        false)) {
+      logger().error("{} onode={} oi too long!");
+      return crimson::ct_error::input_output_error::make();
+    }
+    layout.oi_size = pattr->second.length();
+    pattr->second.copy_out(0, pattr->second.length(), &layout.oi[0]);
```

Comment:
Worth noting here that the actual size of `onode_layout_t` will not fit to the dynamic OI size, which will always be `sizeof(onode_layout_t)`, see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/fltree_onode_manager.cc#L46.

We can only adjust the size after implement `Value::trim()` and `Value::extend()`, see https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/value.h#L191-L195.

Reply:
A fixed size will suffice for now.  Can we use large leaf nodes for now (64?)?

Reply:
Maybe I can store OI_ATTR and SS_ATTR in other extent(s) and add a pointer in onode_layout_t, so that the onode tree won't need to be adjusted.

Reply:
Hmm.  The thing is that every single osd object access hits these bits of metadata.  I don't think it's actually more efficient to move them to a secondary location.  However, whatever is most expedient will do for now -- we can improve it later.

Reply:
I think we can, but we might need to reduce the leaf node size in the future after oid is excluded from value. Larger node means more memcpy and disk I/O for most of tree operations.

----

IIUC TransactionManager already supports 64K extent managements, so the major work would be in the tree implementation.

I think at least we need to remove the hard-coded node size [1], set it to 64K somewhere else and see whether it works immediately. I don't see any size-specific logic that won't work after the change, but very likely there are unexpected corners.

Leaf node split tests [2] can be bothering, because each test layout is manually constructed based on the 4K leaf node to validate 45 different split scenarios. I'm hesitate to rewrite them under 64K, maybe we can just leave these tests with 4K leaf nodes.

There're a number of internal changes pending in the erase/merge feature [3]. It would be better to implement 64K leaf after that. But I think I can reprioritize if needed.

[1] https://github.com/ceph/ceph/blob/master/src/crimson/os/seastore/onode_manager/staged-fltree/stages/node_stage_layout.h#L133
[2] https://github.com/ceph/ceph/blob/master/src/test/crimson/seastore/onode_tree/test_staged_fltree.cc#L513
[3] https://github.com/cyx1231st/ceph/commits/wip-seastore-onode-tree-delete